### PR TITLE
Add Enum support

### LINKED ISSUE #152: Enum support
For now the enum type is no support

### PR DESCRIPTION
**Pull request by ivanmonteiro**
Add enum support to entities

Fix #152

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by ivanmonteiro** [comment]
The Java generator saves enum as string on the database, but dotnet EF saves it as integer by default.
I added custom mappings to save enum as string but it is not being generated.
Check generators/entity-server/templates/dotnetcore/src/Project/Data/ApplicationDatabaseContext.cs.ejs file.
So this PR as draft because of that.

**Comment by ivanmonteiro** [comment]
I can't figure out how to access field.fieldIsEnum inside the template ApplicationDatabaseContext.cs.ejs 
It seems to not be present there, but on other templates I can access it (EntityResourceIntTest.cs.ejs and others).

**Comment by nicolas63** [comment]
I'll look into it as soon as I can.

**Comment by nicolas63** [comment]
entity.definition.fields don't contains fieldIsEnum 

You can use this who work for me 
```
 <%_ entities.forEach(entity => {
                fields.forEach(field => {
                    if (entity.name == entityClass && fields[idx].fieldIsEnum){ _%>
            builder.Entity<<%= entity.name %>>()
                .Property(e => e.<%= fields[idx].fieldNamePascalized %>)
                .HasConversion<string>();
                <%_ }
                });
```


Your code work with jdl but dut don't forget to uncomment enum in entity/prompts.js line 538 for ask enum type with cli

**Comment by ivanmonteiro** [comment]
> entity.definition.fields don't contains fieldIsEnum
> 
> You can use this who work for me
> 
> ```
>  <%_ entities.forEach(entity => {
>                 fields.forEach(field => {
>                     if (entity.name == entityClass && fields[idx].fieldIsEnum){ _%>
>             builder.Entity<<%= entity.name %>>()
>                 .Property(e => e.<%= fields[idx].fieldNamePascalized %>)
>                 .HasConversion<string>();
>                 <%_ }
>                 });
> ```
> 
> Your code work with jdl but dut don't forget to uncomment enum in entity/prompts.js line 538 for ask enum type with cli

Thanks for the suggestions @nicolas63. I will have a look at entity/prompts.js.

I tried your code and it generates the mapping twice

**Comment by nicolas63** [comment]
What is generated twice ? Can you give me a jdl ?

**Comment by ivanmonteiro** [comment]
@nicolas63 the mapping to save the enum as string is generated twice:

            builder.Entity<Proposal>()
                .Property(e => e.Status)
                .HasConversion<string>();
            builder.Entity<Proposal>()
                .Property(e => e.Status)
                .HasConversion<string>();


JDL:

entity Product {
	name String
}

entity Proposal {
	details String
	status ProposalStatus
}

enum ProposalStatus {
    IN_PROGRESS, FINISHED
}

relationship ManyToOne {
	Proposal{product(name)} to Product
}

**Comment by nicolas63** [comment]
oh yes is a bad copy paste ... it's my bad 

try this : 
```
 <%_ entities.forEach(entity => {
                fields.forEach(field => {
                    if (entity.name == entityClass && field.fieldIsEnum){ _%>
            builder.Entity<<%= entity.name %>>()
                .Property(e => e.<%= field.fieldNamePascalized %>)
                .HasConversion<string>();
                <%_ }
                });
```

**Comment by ivanmonteiro** [comment]
> oh yes is a bad copy paste ... it's my bad
> 
> try this :
> 
> ```
>  <%_ entities.forEach(entity => {
>                 fields.forEach(field => {
>                     if (entity.name == entityClass && field.fieldIsEnum){ _%>
>             builder.Entity<<%= entity.name %>>()
>                 .Property(e => e.<%= field.fieldNamePascalized %>)
>                 .HasConversion<string>();
>                 <%_ }
>                 });
> ```

Your code worked 👍 

Now enum is generated using jdl and I'm trying to make it work on the cli

**Comment by ivanmonteiro** [comment]
Main generator v6.10.0 fixed enums not asking for type (see issue: https://github.com/jhipster/generator-jhipster/issues/11898).
Now enum generation should be working with the cli and jdl.

**Comment by nicolas63** [comment]
Nice work ! :smiley: 
Can you add test on the generator for enumeration generation ? 
Something like this https://github.com/jhipster/generator-jhipster/blob/master/test/client.spec.js

Just check if enum is created

**Comment by ivanmonteiro** [comment]
> Nice work ! 😃
> Can you add test on the generator for enumeration generation ?
> Something like this https://github.com/jhipster/generator-jhipster/blob/master/test/client.spec.js
> 
> Just check if enum is created

I have added tests to check enum generation and its mappings at ApplicationDbContext.cs

File: generators/utils.js [hunk 1]
```diff
@@ -18,6 +18,8 @@
  */
 const _ = require('lodash');
 const toPascalCase = require('to-pascal-case');
+const buildEnumInfo = require('generator-jhipster/generators/utils').buildEnumInfo;
```
Comment (nicolas63) [inline generators/utils.js:21]: buildEnumInfo is depracated can you use getEnumInfo ?

Comment (ivanmonteiro) [reply generators/utils.js]: I just changed. Still waiting for CI checks to complete

**Comment by nicolas63** [comment]
Everything seems ok, thank you very much for your contribution there is a bug bounty on this issue don't hesitate to ask for it, you deserved it !

**Pull request merged**